### PR TITLE
PLAT-102257: Migrate Skinnable to hooks

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [x] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Migrates Skinnable to use hooks

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)
This is the first hook (I believe) to expose a context provider. I've gone with `provideSkins` as a member of the returned object from the hook. I like that this avoids exporting the context and doesn't require the user to assemble the provider but it's still a little awkward looking

```jsx
const skins = useSkins(config);
return skins.provideSkins(
    <Component {...props} />
);
```

This model works okay with stacking multiple providers too
```
skins.provideSkins(
    other.provideOther(
        <Comp />
    )
)
```

I've also adjusted the config parameter names a bit for consistency. We had `skins` and `allowedVariants` but both referred to the allowed values. I chose to use `skins` and `variants` rather than `allowedSkins` and `allowedVariants`. I'm not :100: on this, however, since the config also accepts the current values in `skin` and `skinVariants` so we have `skins`/`skin` and `variants`/`skinVariants`.

### CODE REVIEW

File: packages/ui/Skinnable/useSkins.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * 	<Section>
+ * 	<Popup skin="light">
+ * 		<Button>White Button</Button>
+ * 	</Popup>
+ * </App>
+ * ```
+ *
+ * @class SkinContext
+ * @memberof ui/Skinnable
+ * @hoc
```
Comment (webOS101) [inline packages/ui/Skinnable/useSkins.js:22]: Is this really a hoc?  Is it a class?  I think we need a different thing for this.

File: packages/ui/Skinnable/useSkins.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ */
+const SkinContext = React.createContext(null);
+
+/**
+ * Configuration for `useSkins`
+ *
+ * @typedef {Object} useSkinsConfig
+ * @memberof ui/Skinnable
+ * @property {String}   defaultSkin     Default skin name when none has been specified or inherited.
+ * @property {String[]} defaultVariants Default variants when none have been specified or inherited.
```
Comment (webOS101) [inline packages/ui/Skinnable/useSkins.js:33]: is `defaultVariants` required?

File: packages/ui/Skinnable/useSkins.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ * @property {String}   [skin]          Specific skin to apply to this instance which will take precedence over the default or the inherited value.
+ * @property {String[]} skins           List of allowed skins
+ * @property {String[]} [skinVariants]  Specific variants to apply to this instance which will take precedence over the default or the inherited value.
+ * @property {String[]} variants        List of allowed variants
```
Comment (webOS101) [inline packages/ui/Skinnable/useSkins.js:37]: Similarly, is `variants` actually required?

File: packages/ui/Skinnable/useSkins.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ *     defaultSkin: 'neutral',
+ *     defaultVariants: [],
+ *     skins: ['neutral', 'bold'],
+ *     variants: ['highContrast'],
+ *     skin: props.skin,
+ *     skinVariants: props.skinVariants
+ *   });
+ *
+ *   return skins.provideSkins(
+ *     <Component className={classnames(props.className, skins.className)} />
```
Comment (webOS101) [inline packages/ui/Skinnable/useSkins.js:74]: Would this be clearer showing `provideSkins` and `className` destructured from the call?  Along the same vein, can we provide a utility function from `useSkins` so developers don't have to call `classnames`?  Maybe return a utility `mergeSkinClasses()`?

File: packages/ui/Skinnable/util.js [hunk 5]
```diff
@@ -26,7 +27,58 @@ const objectify = (arg) => {
 
 const preferDefined = (a, b) => ((a != null) ? a : b);
 
+function determineSkin (defaultSkin, authorSkin, parentSkin) {
+	return authorSkin || defaultSkin || parentSkin;
```
Comment (webOS101) [inline packages/ui/Skinnable/util.js:31]: `defaultSkin` here is the local default, not the app default.  I'm not sure if there's a way to clarify that or if it needs to, but I was confused at first.

File: packages/ui/Skinnable/useSkins.js (hunk 1 above)
Comment (ryanjduffy) [reply packages/ui/Skinnable/useSkins.js:22]: It's private so I just removed the pragma.

File: packages/ui/Skinnable/useSkins.js (hunk 2 above)
Comment (ryanjduffy) [reply packages/ui/Skinnable/useSkins.js:33]: Nope

File: packages/ui/Skinnable/useSkins.js (hunk 3 above)
Comment (ryanjduffy) [reply packages/ui/Skinnable/useSkins.js:37]: Nope

File: packages/ui/Skinnable/util.js (hunk 5 above)
Comment (ryanjduffy) [reply packages/ui/Skinnable/util.js:31]: Adding some docs for this method and the one below.

File: packages/ui/Skinnable/useSkins.js (hunk 4 above)
Comment (ryanjduffy) [reply packages/ui/Skinnable/useSkins.js:74]: I can see the utility of providing that method but I think it goes beyond the scope of this feature. Most likely, consumers will need other CSS class concatenation and would rather not introduce overlapping capabilities in the hooks.

Sound reasonable?

Comment (webOS101) [reply packages/ui/Skinnable/useSkins.js:73]: Seems reasonable!

**Pull request merged**